grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+use std::{fs::File, io::Read, path::Path, time::Duration};
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+
+use genos::{
+    output::{Content, Output, RichTextMaker, Section, StatusUpdates, Update},
+    points::PointQuantity,
+    process::{Command, ExitStatus, ProcessExecutor, SignalType, StdinPipe},
+    stage::{StageResult, StageStatus},
+    Executor,
+};
+
+use regex::Regex;
+
+use serde::Deserialize;
+
+use tracing::debug;
+
+// default should be longer than run stage default
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(120);
+
+// exit code used to identify non-signal failures.
+// not configurable via YAML.
+//
+// exit codes >125 are reserved by POSIX; see exit(1p)
+// and https://tldp.org/LDP/abs/html/exitcodes.html
+const ERROR_EXITCODE: i32 = 125;
+
+#[derive(Debug, Default, Deserialize, Clone)]
+pub struct ValgrindConfig {
+    log_file: String,
+    leak_check: Option<bool>,
+    malloc_fill: Option<u8>,
+    free_fill: Option<u8>,
+    suppressions: Option<Vec<String>>,
+}
+
+pub struct Valgrind<E> {
+    executor: E,
+    config: ValgrindConfig,
+    executable: String,
+    args: Vec<String>,
+    stdin: Option<String>,
+    timeout: Option<Duration>,
+}
+
+fn read_file(path: &Path) -> Result<String> {
```

Comment:
NP: Could make this async and use the async versions of File::open and read_to_string found in tokio. 
Also this pattern is pretty common, maybe we could move this to genos::fs as a function available everywhere?

Reply:
Definitely should be a library method. I can move it to genos::fs and make it async